### DISCUSSION TITLE
Images not being loaded

### DISCUSSION THREAD

**Discussion by syntafin** [pixelfed/pixelfed#5138] [Support]
Good day,

installed pixelfed today, and have some problems now.
As you can see here:
https://eorzeadates.com/i/web/post/499341758367911937

There is the "no preview" image loaded, As in ACP noted this would be the image:
https://eorzeadates.com/storage/m/_v2/499319305660010497/a48ad3dc0-89cf68/K01cEOvv5CLh/CGJ1NzREXLimdZV4crYHKLfqQ4rkihdkvoLKVUn0.png

What isnt there too... Storage is linked and the file with this name is there:
![image](https://user-images.githubusercontent.com/695535/202820582-33f6522d-e1e4-493c-a0b2-71a59eaf8447.png)
(ignore queue worker there)

I use Horizon for the queue and cached routes/views/config. Anyone know how to solve it?

**Comment by mraiur** [comment]
You can check the nginx log and the laravel log when you hit the url to see what the error is.
nginx : /var/log/nginx/error.log
laralel : storage/log/laravel.log

**Comment by syntafin** [comment]
And there is something more strange: Permission denied:
![Screenshot_20221119-104849.png](https://user-images.githubusercontent.com/695535/202845058-297745c3-0f6c-4970-8bd8-3c3770820ddd.png)

First I thought it was caused by the fact that I started Horizon under the wrong user, but even now when it runs with the correct user, it throws that error message.

**Comment by syntafin** [comment]
Was able to fix it, when I disable isolation mode and run it on default user.

**Comment by shanesipe** [comment]
how do you disable isolation mode

**Comment by syntafin** [comment]
> how do you disable isolation mode

![image](https://user-images.githubusercontent.com/695535/202866642-727781ac-94c4-4386-a173-b55ffff9293d.png)

I use Laravel Forge to deploy my apps and provision my servers.
There you can use a "Website Isolation" what runs FPM, NGINX under a isolated user instead of the main user.

**Comment by shanesipe** [comment]
I am on cli with ubuntu so doesnt help thanks though

**Comment by syntafin** [comment]
you have the same problem?

**Comment by shanesipe** [comment]
yes cant access storage and everything else have changed permissions nothing helps

**Comment by shanesipe** [comment]
22/11/19 17:53:58 [crit] 119495#119495: *2843 stat() "/var/www/pixelfed/public/storage/m/_v2/499448784272666625/a48ad3dc0-89cf68/dtWdCzn1DJnJ/74sNVrZtxhYFBIAYBl6VIOvkvwcizbSagxkvt0vX.jpg" failed (13: Permission denied), client: 64.33.224.57, server: test.domain.com, request: "GET /storage/m/_v2/499448784272666625/a48ad3dc0-89cf68/dtWdCzn1DJnJ/74sNVrZtxhYFBIAYBl6VIOvkvwcizbSagxkvt0vX.jpg HTTP/2.0", host: "test.domain.com", referrer: "https://test.domain.com/i/web"
2022/11/19 17:53:58 [crit] 119495#119495: *2843 stat() "/var/www/pixelfed/public/storage/m/_v2/499448784272666625/a48ad3dc0-89cf68/dtWdCzn1DJnJ/74sNVrZtxhYFBIAYBl6VIOvkvwcizbSagxkvt0vX.jpg" failed (13: Permission denied), client: 64.33.224.57, server: test.domain.com, request: "GET /storage/m/_v2/499448784272666625/a48ad3dc0-89cf68/dtWdCzn1DJnJ/74sNVrZtxhYFBIAYBl6VIOvkvwcizbSagxkvt0vX.jpg HTTP/2.0", host: "test.domain.com", referrer: "https://test.domain.com/i/web"

**Comment by syntafin** [comment]
How you process images? Normal queue or horizon?

**Comment by shanesipe** [comment]
> How you process images? Normal queue or horizon?

I assume horizon not sure though

**Comment by creideiki** [comment]
I think I'm seeing the same thing on the instance I'm in the process of setting up. When trying to make my first post, the images were uploaded successfully, but the web UI doesn't show them when asking for captions and tags. Network debugging in the web browser shows "403 Forbidden" for a lot of requests under `/storage/m/_v2/$RANDOM/`. Web server logs say: `$ROOT/pixelfed/public/storage/m/_v2/.htaccess pcfg_openfile: unable to check htaccess file, ensure it is readable and that '$ROOT/pixelfed/public/storage/m/_v2/' is executable`.

Which it is not:
```
pixelfed@pixelfed ~/public/storage/m/_v2 $ ls -la
total 0
drwx------ 3 pixelfed pixelfed 40 Nov 20 18:22 .
drwxr-xr-x 3 pixelfed pixelfed 47 Nov 20 18:22 ..
drwx------ 3 pixelfed pixelfed 38 Nov 20 18:22 499972618752536577
```
```
pixelfed@pixelfed ~/public/storage/m/_v2/499972618752536577 $ ls -l
total 0
drwx------ 7 pixelfed pixelfed 130 Nov 20 18:22 a48ad3dc0-89cf68
```
It looks like the `public/storage/m/_v2/` directory was created by the PHP-FPM process, running as the user `pixelfed`, with access bits set to `700`. Then, my Apache server tries to read the static files as the user `apache`, which it is not allowed to do.

If I allow the world to access those files:
```
pixelfed@pixelfed ~/public/storage/m $ chmod -R o+r _v2/
pixelfed@pixelfed ~/public/storage/m $ find _v2/ -type d -exec chmod o+x '{}' ';'
```
Then the images appear in the web UI. So it seems that the only error is that they are created with the wrong access bits.

**Comment by syntafin** [comment]
> > How you process images? Normal queue or horizon?
> 
> I assume horizon not sure though

how you can't be sure? you started the daemon?

@creideiki idk how it is for Apache2, but may you run FPM and Apache2 under the same user.
If I find the time, I will check was is going wrong bc it doesnt work as expected.

FYI I have a similar app developed myself, without this problem (but even without ActivityPub integration).

**Comment by okpierre** [comment]
Maybe this will fix it @syntafin @creideiki @shanesipe @mraiur [3825](https://github.com/pixelfed/pixelfed/pull/3825)

**Comment by syntafin** [comment]
> Maybe this will fix it @syntafin @creideiki @shanesipe @mraiur [3825](https://github.com/pixelfed/pixelfed/pull/3825)

Hope @dansup is back soon 😅 I got some improvements in the pipeline, but need to reset my fork.... Don't ask why, I was stupid.
But PR looks good, I thought it would be something in that way, but hadn't the time to check it there.

**Comment by creideiki** [comment]
I wanted to keep my installation setup, so didn't try moving everything to the Apache user. I have instead tried running with the patch in #3825, and it seems to fix my problem with images not being available directly to the web server. I don't know enough PHP to say if it introduces any other problems, though.

**Comment by fofwisdom** [comment]
> I am on cli with ubuntu so doesnt help thanks though

I have same problem.

Ubuntu 23.04
Nginx 1.22.0
php 8.1-fpm
Redis 7.0.8

Composer 2.57
Pixelfed 0.11.8
Horizon service is fine.


Using dedicated user pixelfed.
The owner and group of the Pixelfed path is pixelfed:pixelfed, with permissions of 755/644.
![image](https://github.com/pixelfed/pixelfed/assets/205521/bf00eee5-a1da-4daa-8b0e-aa52a6774f22)
pixelfed belongs to the www-data and redis groups.
The user and group in php-fpm is pixelfed.
![image](https://github.com/pixelfed/pixelfed/assets/205521/959f01a1-b112-486a-89f6-ad09a2f520db)
The user in the Horizon service is pixelfed.
![image](https://github.com/pixelfed/pixelfed/assets/205521/7db7172b-169b-41ec-9441-23fee01db65c)


Turning on APP_DEBUG doesn't show any errors in the webpage
Laravel does not display any errors.
Nginx says there are no permissions.
> 2023/06/04 16:05:43 [crit] 60297#60297: *210 stat() "/home/pixelfed/live/public/storage/m/_v2/570153857971580929/186f813ba-297a89/eU45Avpr6Ukw/aCS4Y8oYvHLiUB8fvokpLSfmnaIzKFwl75wDH4Fp.png" failed (13: Permission denied), client: 192.168.146.101, server:
> gal.muri.network, request: "GET /storage/m/_v2/570153857971580929/186f813ba-297a89/eU45Avpr6Ukw/aCS4Y8oYvHLiUB8fvokpLSfmnaIzKFwl75wDH4Fp.png?v=1685862342 HTTP/2.0", host: "gal.muri.network", referrer: "https://gal.muri.network/i/web/compose"


Changing the permissions of the image to 777 does not allow access to the image.
![image](https://github.com/pixelfed/pixelfed/assets/205521/57871fcc-fdaf-4c12-bc97-3fcabe88eeff)
![image](https://github.com/pixelfed/pixelfed/assets/205521/862380bb-b151-4ecb-983b-860942e12c0b)

**Comment by syntafin** [comment]
What are the permission of the folder?

**Comment by fofwisdom** [comment]
### Permission
user pixelfed
group pixelfed
Directories 755
files 644

### Solved problem
The patch in commit [22da264](https://github.com/pixelfed/pixelfed/commit/22da2647c7f28f7e1d6a588f4549c3326aee9356#diff-e1f5f65898d53a762d4676492d99c81863e1e30f61a1af2c06d0acccf28c513f) gave less permissions to private. The file is not accessible via dedicated user.

Replace the code as shown below and the issue is resolved.
```
            'permissions' => [
                'file' => [
                    'public' => 0644,
                    'private' => 0640,
                ],
                'dir' => [
                    'public' => 0755,
                    'private' => 0750,
```
replaced the private entries from 0600 and 0700 to 0640 and 0750.

Thanks.
